File: C++/021. Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
		ListNode result(-1);
		for (ListNode* it = &result; l1 != NULL || l2 != NULL; it = it->next) {
		    if (l1 == NULL)
		    {
		        it->next = l2;
		        break;
		    }
		    if (l2 == NULL)
		    {
		        it->next = l1;
		        break;
		    }
			if (l1->val <= l2->val) {
				it->next = l1;
				l1 = l1->next;
			}
			else {
				it->next = l2;
				l2 = l2->next;
			}
		}
		return result.next;
	}
};

File: python/021. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head= ListNode(0)
        prev = head
        prev.next = l1
        node1, node2 = l1, l2
        while node1 and node2:
            if node2.val < node1.val:
                temp = node2
                node2 = node2.next
                prev.next = temp
                prev.next.next = node1
                prev = prev.next
            else:
                node1 = node1.next
                prev = prev.next
        if node2: prev.next = node2
        return head.next

